apf/apfAggregateNumbering.cc

```diff
@@ -0,0 +1,365 @@
+    //  are reset in globalize which must be called after
+    //  init for the numbering to be valid
+    strds.resize(static_cast<unsigned>(sz));
+    // all peers in the aggregation scope necessarily
+    //  have the same stride
+    int strd = PCU_Add_Int(lcl_strd);
+    for(int ii = 0; ii < sz; ++ii)
+      strds[ii] = strd;
+    // the first dof id of the first local node is the # of
+    //  nodes before the first local node times the # of
+    //  dofs in the first block on a node
+    int dof_id_offset = lcl_strd * dofs_per_blk;
+    int frst = PCU_Exscan_Int(dof_id_offset);
+    frsts.resize(static_cast<unsigned>(sz));
+    MPI_Allgather(&frst,1,MPI_INTEGER,&frsts[0],1,MPI_INTEGER,PCU_Get_Comm());
+    T nd_id = 0;
+    for(int dd = 0; dd < dim; ++dd)
+    {
+      MeshIterator * it = m->begin(dd);
+      MeshEntity * e = NULL;
+      while((e = m->iterate(it)))
+      {
+        int nds = FieldBase::shape->countNodesOn(m->getType(e));
+        // only really works if nds == 1 or 0
+        //  due to the above comment ^
+        for(int nd = 0; nd < nds; ++nd)
+        {
+          if(shr->isOwned(e))
+          {
+            nd_ids->set(e,&nd_id);
+            ++nd_id;
+          }
+        }
+      }
+      m->end(it);
+    }
+  }
+  template <class T>
+  void AggregateNumberingOf<T>::init(Field * fld,
+                                     Sharing * sh,
+                                     int blks,
+                                     int bs)
+  {
+    PCU_ALWAYS_ASSERT(fld->countComponents() == blks * bs);
+    std::string nm = fld->getName();
+    nm += "_num";
+    init(nm.c_str(),
+         fld->getMesh(),
+         sh,
+         fld->getShape(),
+         blks,
+         bs);
+  }
+  template <class T>
+  void AggregateNumberingOf<T>::globalize()
+  {
+    MPI_Comm gbl_cm = PCU_Get_Comm();
+    synchronizeFieldData(nd_ids,shr,false);
+    // any rank with frst == 0 was the zero rank
+    //  in the aggregation scope under which init
+    //  was called, the offsets from these
+    //  zero-ranked processes are what we need to
+    //  sum since they include in their strides
+    //  all the nodes from the local aggregation scope
+    int strd = getScopeStride();
+    int frst = getLocalFirstDof();
+    int scp_offset = (int)(!(bool)frst) * (strd * dofs_per_blk * blks_per_nd);
+    int lcl_offset = PCU_Exscan_Int(scp_offset);
+    lcl_offset += scp_offset; // make the excan a scan
+    // remove the contribution from the local
+    //  aggregation scope, because the scope can be
+    //  any comm and only rank 0 in each scope
+    //  contributes to the above scan
+    lcl_offset -= (strd * dofs_per_blk * blks_per_nd);
+    frst += lcl_offset;
+    // strds and frsts needs to be resized/updated
+    //  since we can have adjacent parts that were
+    //  not in the aggregation scope under which
+    //  init was called
+    int sz = PCU_Comm_Peers();
+    frsts.setSize(sz);
+    MPI_Allgather(&frst,1,MPI_INTEGER,
```

Comment:
I don't recall if PCU provides `Allgather`.  If not, we should consider adding it to avoid/minimize direct MPI calls.  (repeat for the other `Allgather` calls)